docs/Software Engineering Blog/posts/spring-boot-3-1-docker-compose-support-intuitive-consistent-develop-run-tests-containers.md

```diff
@@ -0,0 +1,229 @@
+cluster.
+
+## Little More Setup to Make It Better
+
+To enhance the development and testing experience, consider the following improvements.
+
+### For Development
+
+#### Docker Profile Connected to Spring Profile:
+
+Since the practice of spring profiles is usually used as a dependency management method. For example, we add a
+profile `redis` in `spring.profiles.include`, to include `application-redis.yaml`, which contains the related config for
+using redis.
+
+So we want to link docker compose profiles to Spring profiles. In this way, developers can ensure that the correct
+configuration is used and the corresponding container is up also.
+
+To do that, we need some modification:
+
+1. use comma separate format for `spring.profiles.include`
+
+  ```yaml title="application.yaml"
+  spring:
+    application:
+      name: order-command-side
+    profiles:
+      include: "observe, cluster, mock, kafka, kafka-producer, redis, idempotency, docker-support"
+  ```
+
+2. add profile linking property 'spring.docker.compose.profiles.active'
+
+  ```yaml title="application-docker-support.yaml"
+  spring:
+    docker:
+      compose:
+        enabled: true
+        profiles:
+          active: ${spring.profiles.include}
+  ```
+
+3. add profile for each service in compose.yaml
+
+  ```yaml title="compose.yaml"
+  version: "3"
+
+  services:
+    kafka:
+      image: 'bitnami/kafka:3.7'
+      profiles: ["kafka", "all"]
+      ports:
+        - '9092:9092'
+      environment:
+        - KAFKA_CFG_NODE_ID=0
+        - KAFKA_CFG_PROCESS_ROLES=controller,broker
+        - KAFKA_CFG_LISTENERS=PLAINTEXT://:9092,CONTROLLER://:9093
+        - KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT
+        - KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://${KAFKA_HOST:-localhost}:9092
+        - KAFKA_CFG_CONTROLLER_QUORUM_VOTERS=0@kafka:9093
+        - KAFKA_CFG_CONTROLLER_LISTENER_NAMES=CONTROLLER
+
+    redis:
+      image: 'redis:7.2'
+      profiles: ["redis", "all"]
+      ports:
+        - '6379:6379'
+  ```
+
+#### Ignore when Building in Image:
+
+[JIB](https://github.com/GoogleContainerTools/jib) is now an out-of-the-box technic of building image for a Spring Boot Application. For now, JIB still include the spring-boot-docker-compose library in the built image (we can track on [this issue](https://github.com/GoogleContainerTools/jib-extensions/issues/158) for new solutions). However, in most cases, we have existed dependency components (e.g. DB service, Redis cluster...) in Test, Staging, or Production environment. So we need to manually exclude it by setting up by jib-layer-filter-extension for [gradle](https://github.com/GoogleContainerTools/jib-extensions/tree/master/first-party/jib-layer-filter-extension-gradle)/[maven](https://github.com/GoogleContainerTools/jib-extensions/tree/master/first-party/jib-layer-filter-extension-maven).
```

Comment:
📝 **[vale]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Google.Contractions] Use 'we've' instead of 'we have'.